pkg/cloud/services/ec2/launchtemplate_test.go

```diff
@@ -309,3 +436,850 @@ func TestService_LaunchTemplateNeedsUpdate(t *testing.T) {
+				Client:     client,
+				Cluster:    newCluster(),
+				AWSCluster: newAWSCluster(),
+			})
+			if err != nil {
+				t.Fatalf("Failed to create test context: %v", err)
+			}
+
+			ms, err := scope.NewMachinePoolScope(scope.MachinePoolScopeParams{
+				Client:         client,
+				InfraCluster:   cs,
+				Cluster:        newCluster(),
+				MachinePool:    newMachinePool(),
+				AWSMachinePool: newAWSMachinePool(),
+			})
+			if err != nil {
+				t.Fatalf("Failed to create test context: %v", err)
+			}
+			ms.AWSMachinePool.Spec.AWSLaunchTemplate = tc.awsLaunchTemplate
+			ms.MachinePool.Spec.Template = tc.machineTemplate
+
+			tc.expect(ec2Mock.EXPECT())
+
+			s := NewService(cs)
+			s.EC2Client = ec2Mock
+
+			id, err := s.DiscoverLaunchTemplateAMI(ms)
+			tc.check(id, err)
+		})
+	}
+}
+
+func TestDeleteLaunchTemplateVersion(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	type args struct {
+		id      string
+		version *int64
+	}
+	testCases := []struct {
+		name   string
+		args   args
+		expect func(m *mock_ec2iface.MockEC2APIMockRecorder)
+		check  func(err error)
+	}{
+		{
+			name:   "version is a nil pointer",
+			expect: func(m *mock_ec2iface.MockEC2APIMockRecorder) {},
+			check: func(err error) {
+				if err == nil {
+					t.Fatalf("expected an error")
+				}
+			},
+		},
+		{
+			name: "delete launch template call fails from AWS",
+			args: args{
+				id:      "id",
+				version: aws.Int64(12),
+			},
+			expect: func(m *mock_ec2iface.MockEC2APIMockRecorder) {
+				m.DeleteLaunchTemplateVersions(gomock.Eq(
+					&ec2.DeleteLaunchTemplateVersionsInput{
+						LaunchTemplateId: aws.String("id"),
+						Versions:         aws.StringSlice([]string{"12"}),
+					},
+				)).Return(nil, awserrors.NewConflict("Some issue with AWS"))
+			},
+			check: func(err error) {
+				if err == nil {
+					t.Fatalf("expected an error")
+				}
+			},
+		},
+		{
+			name: "success: delete launch template version",
+			args: args{
+				id:      "id",
+				version: aws.Int64(12),
+			},
+			expect: func(m *mock_ec2iface.MockEC2APIMockRecorder) {
+				m.DeleteLaunchTemplateVersions(gomock.Eq(
+					&ec2.DeleteLaunchTemplateVersionsInput{
+						LaunchTemplateId: aws.String("id"),
+						Versions:         aws.StringSlice([]string{"12"}),
+					},
+				)).Return(nil, nil)
+			},
+			check: func(err error) {
+				if err != nil {
+					t.Fatalf("did not expect an erro, but got %+v", err)
+				}
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			ec2Mock := mock_ec2iface.NewMockEC2API(mockCtrl)
+
+			scheme := runtime.NewScheme()
+			_ = infrav1.AddToScheme(scheme)
+			client := fake.NewClientBuilder().WithScheme(scheme).Build()
+
+			scope, err := scope.NewClusterScope(scope.ClusterScopeParams{
+				Client:     client,
+				Cluster:    &clusterv1.Cluster{},
+				AWSCluster: &infrav1.AWSCluster{},
+			})
+			if err != nil {
```

Comment:
what do you think about using `gomega` matchers? like here https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/50af6ec5ce7a183b67d23a187e687c42fef24176/pkg/cloud/services/iamauth/configmap_test.go#L150

Reply:
done